File: c/puissance4.c
#include <stdio.h>

int x=6;
int y=6;
int player = 0;
char *playersym[2] = {"x\0", "o\0"}; 
char *matrix[6][6];
int nbjeton = 36;
int winner = 0;
int col  = 0;

int check_dir(int dx,int dy, int ligne, int col){
	int result=0;
	int px=ligne;
	int py=col;
	px+=dx;
	py+=dy;
	while ( (px>=0 && px<x) && (py>=0 && py<y) && (matrix[px][py] == matrix[ligne][col]) ){
		result++;
		px+=dx;
		py+=dy;
	}
	return result;
}

int checkwinner(int ligne, int col){
	return ( ((check_dir(-1, 0, ligne, col)+check_dir(1, 0, ligne, col)+1)>=4) || ((check_dir(0, 1, ligne, col)+check_dir(0, -1, ligne, col)+1)>=4) || ((check_dir(1, -1, ligne, col)+check_dir(-1, 1, ligne, col)+1)>=4) );
}

void initgame(){
	for(int i=0; i<x;i++){
		for(int v=0; v<y;v++){ 
			matrix[i][v]=" \0";
		}
	}
}

void showgame(){
	printf("Tour du joueur : %s\n\n", playersym[player]);
	printf("| 0 | 1 | 2 | 3 | 4 | 5 |\n-------------------------\n");
	for(int i=0; i<x;i++){
		for(int v=0; v<y;v++){
			printf("| %s ",matrix[i][v]);
		}
		printf("|\n-------------------------\n");
	}
}

int playgame(){
	int ligne = 0;
	printf("Please choose a colum : ");
	scanf("%d", &col);
	col%=x;
	while( matrix[ligne][col] == " \0" && x>ligne){
		ligne++;
	}
	matrix[ligne-1][col]=playersym[player];
	return ligne-1;
}

void setplayer(){
	player=(player+1)%2;
}

int  main(){
	initgame();
	while (!winner && nbjeton){
		printf("\033[H\033[J");
		showgame();
		int ligne = playgame();
		winner = checkwinner(ligne, col);
		nbjeton--;
		setplayer();
	}
	showgame();
}

File: python/puissance4.py
x=6;
y=6;
player = 0;
playersym = ["x", "o"]
matrix = [[0]*(x+1) for i in range(y+1)]
nbjeton = 36
winner = False
col  = 0

def check_dir(dx, dy, ligne, col):
    result=0
    px=ligne
    py=col
    px+=dx
    py+=dy
    while px>=0 and px<x and py>=0 and py<y and matrix[px][py] == matrix[ligne][col]:
        result+=1
        px+=dx
        py+=dy
    return result

def checkwinner(ligne, col):
    return (check_dir(-1, 0, ligne, col)+check_dir(1, 0, ligne, col)+1)>=4 or (check_dir(0, 1, ligne, col)+check_dir(0, -1, ligne, col)+1)>=4 or (check_dir(1, -1, ligne, col)+check_dir(-1, 1, ligne, col)+1)>=4

def initgame():
    for i in range(0, x+1):
        for v in range(0, y+1):
            matrix[i][v]=" "


def showgame():
    print("Tour du joueur :" + playersym[player])
    print("| 0 | 1 | 2 | 3 | 4 | 5 |\n-------------------------")
    for i in range(x):
        for v in range(y):
            print("| " +  matrix[i][v], end=" ")
        print("|\n-------------------------")

def playgame():
    ligne = 0
    global col 
    getint()
    #col = int(input('Please choose a colum : '))
    col%=x;
    while  matrix[ligne+1][col].isspace() and (x-1)>ligne:
        ligne+=1
    matrix[ligne][col]=playersym[player];
    return ligne

def getint():
    while True:
        try:
            col = int(input('Please choose a colum : '))
            break
        except ValueError:
            print("Try Again")
    

    return
def setplayer():
    global player
    player=(player+1)%2

def maingame():
    global winner
    global nbjeton
    initgame()
    while not winner and nbjeton:
        setplayer()
        print("\033[H\033[J");
        showgame()
        ligne = playgame();
        winner = checkwinner(ligne, col)
        nbjeton-=1

maingame()
print("\033[H\033[J");
print("BRAVO : " + playersym[player] + " Tu as GAGNE !!!!!")
showgame()

